Sources/src/FromSTMTestBeamData_source.cc

```diff
@@ -0,0 +1,335 @@
+    }
+    else if (configuration.find("LaBr") != std::string::npos) {
+      channel_ = STMChannel::LaBr;
+    }
+    else {
+      throw cet::exception("FromSTMTestBeamData") << "Cannot determine the channel from the configuration field (" << configuration << "). This should contain either \"HPGe\" or \"LaBr\"" << std::endl;
+    }
+
+    // Get the run and subrun numbers from the sequencer
+    std::string sequencer = tokens.at(4);
+    std::string runNo = sequencer.substr(0, 6); // sequencer always has 6 characters for run
+    runNumber_ = std::stoi(runNo);
+
+    if(!art::RunID(runNumber_).isValid()) {
+      throw cet::exception("BADCONFIG", " FromSTMTestBeamData: ")
+        << " fhicl::ParameterSet specifies an invalid runNumber = "<<runNumber_<<"\n";
+    }
+
+    if(runNumber_ < 101000 || runNumber_ > 101999) {
+      throw cet::exception("FromSTMTestBeamData") << "Run number is outside of our reserved run range (101000 -- 101999)" << std::endl;
+    }
+
+    // There are two different versions of the binary file:
+    //  - v2 contains a unix timestamp after the trigger header
+    //  - v1 does not
+    // For the time being, just hardcode which runs belong to which (ideally would be in a DB?)
+    if (runNumber_ == 101001) {
+      binaryFileVersion_ = 2;
+    }
+    else if (runNumber_ >= 101002 && runNumber_<=101014) {
+      binaryFileVersion_ = 1;
+    }
+    else {
+      binaryFileVersion_ = 2;
+    }
+
+    std::string subrunNo = sequencer.substr(7, 7+8); // sequencer always has 6 characters for run followed by an underscore and then 8 characters for the sub run
+    currentSubRunNumber_ = std::stoi(subrunNo);
+    currentEventNumber_ = 1;
+    fb = new art::FileBlock(art::FileFormatVersion(1, "STMTestBeamDataInput"), currentFileName_);
+  }
+
+  //----------------------------------------------------------------
+  void STMTestBeamDataDetail::closeCurrentFile() {
+    currentFileName_ = "";
+    currentFile_->close();
+    delete currentFile_;
+    currentFile_ = nullptr;
+  }
+
+  //----------------------------------------------------------------
+  bool STMTestBeamDataDetail::readNext(art::RunPrincipal* const& inR,
+      art::SubRunPrincipal* const& inSR,
+      art::RunPrincipal*& outR,
+      art::SubRunPrincipal*& outSR,
+      art::EventPrincipal*& outE)
+  {
+    if (currentFile_->eof()) {
+      return false;
+    }
+
+    // Create the STMWaveformDigiCollection that we will write to the art event
+    // NB although we are creating one collection for each detector, in the test beam only one
+    //    detector was connected at a time so one of these collections will be empty
+    std::unique_ptr<mu2e::STMWaveformDigiCollection> outputHPGeWaveformDigis(new mu2e::STMWaveformDigiCollection());
+    std::unique_ptr<mu2e::STMWaveformDigiCollection> outputLaBrWaveformDigis(new mu2e::STMWaveformDigiCollection());
+
+    // Read the trigger header
+    STMTestBeam::TriggerHeader trigger_header[1];
```

Comment:
I am curious why you made this an array of length 1 and not just a plain on object of that type.  The code as it exists now would work fine if it were just an object.  Are you preparing for a future development?

Reply:
I think I was just copying an example that used an array of length 1. I have changed to a plain object in [6ceadcb](https://github.com/Mu2e/Offline/pull/856/commits/6ceadcbf6c3dca6beccfc5cb6389cbb5513019fd)